src/VisualStudio/Core/Test.Next/Remote/RemoteHostClientServiceFactoryTests.cs

```diff
@@ -71,6 +73,103 @@ public async Task UpdaterService()
+
+        [Fact]
+        public async Task TestStreamingServices()
+        {
+            using var workspace = CreateWorkspace();
+
+            var exportProvider = workspace.Services.SolutionServices.ExportProvider;
+            var listenerProvider = exportProvider.GetExportedValue<AsynchronousOperationListenerProvider>();
+            var globalOptions = exportProvider.GetExportedValue<IGlobalOptionService>();
+
+            var service = workspace.Services.GetRequiredService<IRemoteHostClientProvider>();
+            using var client = await service.TryGetRemoteHostClientAsync(CancellationToken.None);
+
+            // add solution, change document
+            workspace.AddSolution(SolutionInfo.Create(SolutionId.CreateNewId(), VersionStamp.Default));
+            var project = workspace.AddProject("proj", LanguageNames.CSharp);
+            var document = workspace.AddDocument(project.Id, "doc.cs", SourceText.From(""));
+
+            var assetStorage = workspace.Services.GetRequiredService<ISolutionAssetStorageProvider>().AssetStorage;
+
+            var tasks = new List<Task>();
+
+            // Kick off a ton of work in parallel to try to shake out race conditions with pinning/syncing.
+
+            // 100 outer loops that ensure we generate the same files with the same checksums at least 10 times.
+            for (var i = 0; i < 30; i++)
+            {
+                // 100 inner loops producing variants of the file.
+                for (var j = 0; j < 30; j++)
+                {
+                    tasks.Add(Task.Run(async () => await PerformSearchesAsync(client, document, name: "Goo" + i, i + j)));
+                }
+            }
+
+            await Task.WhenAll(tasks);
+
+            Assert.Equal(0, assetStorage.GetTestAccessor().PinnedScopesCount);
+        }
+
+        private static async Task PerformSearchesAsync(RemoteHostClient client, Document document, string name, int spaces)
+        {
+            // Fork the document, with 100 variants of methods called 'name' in it.
+            var text = SourceText.From(CreateText(name, spaces));
+            var forked = document.Project.Solution.WithDocumentText(document.Id, text);
+
+            // Search teh forked document, ensuring we find all the results we expect.
+            var channel = Channel.CreateUnbounded<RoslynNavigateToItem>();
+            var cancellationToken = CancellationToken.None;
+
+            Task.Run(async () => await client.TryInvokeAsync<IRemoteNavigateToSearchService>(
+                    forked,
+                    (service, solutionChecksum, callbackId, cancellationToken) =>
+                        service.SearchProjectAsync(
+                            solutionChecksum, document.Project.Id, ImmutableArray<DocumentId>.Empty, name, s_kinds, callbackId, cancellationToken),
+                    new NavigateToSearchServiceCallback(channel), cancellationToken).ConfigureAwait(false))
+                .CompletesChannel(channel);
```

Comment:
we're not actually using IAE over OOP anymore.  But i still liked having this test.